x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -727,16 +830,337 @@ public void runAsDenied(String requestId, Authentication authentication, RestReq
+            logEntry.with(DELETE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(DeleteRoleMappingRequest deleteRoleMappingRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "delete_role_mapping");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("role_mapping")
+                    .field("name", deleteRoleMappingRequest.getName())
+                    .endObject() // role_mapping
+                    .endObject();
+            logEntry.with(DELETE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(InvalidateApiKeyRequest invalidateApiKeyRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "invalidate_apikeys");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("apikeys");
+                if (invalidateApiKeyRequest.getIds() != null && invalidateApiKeyRequest.getIds().length > 0) {
+                    builder.array("ids", invalidateApiKeyRequest.getIds());
+                }
+                if (Strings.hasLength(invalidateApiKeyRequest.getName())) {
+                    builder.field("name", invalidateApiKeyRequest.getName());
+                }
+                builder.field("owned_by_authenticated_user", invalidateApiKeyRequest.ownedByAuthenticatedUser());
+                if (Strings.hasLength(invalidateApiKeyRequest.getUserName()) || Strings.hasLength(invalidateApiKeyRequest.getRealmName())) {
+                    builder.startObject("user")
+                            .field("name", invalidateApiKeyRequest.getUserName())
+                            .field("realm", invalidateApiKeyRequest.getRealmName())
+                            .endObject(); // user
+                }
+            builder.endObject() // apikeys
+                    .endObject();
+            logEntry.with(INVALIDATE_API_KEYS_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(DeletePrivilegesRequest deletePrivilegesRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "delete_privileges");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("privileges")
+                        .field("application", deletePrivilegesRequest.application())
+                        .array("privileges", deletePrivilegesRequest.privileges())
+                    .endObject() // privileges
+                    .endObject();
+            logEntry.with(DELETE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
```

Comment:
This PR adds 500 lines to a file that is already over 1,000 lines long.
I really think we need to split it up in some way.
Happy for it to be a separate PR, but this class is getting way too huge.

Reply:
I wholeheartedly agree that we need to invest in restructuring this code. I should've probably seen it coming before this PR. As you say we can tackle in a follow-up.

This class took hold of the log line format and settings to alter that format. I tried to splice out the `LogEntryBuilder` class, and use that on a separate logger, for the new events, but it depends on instance members of the `LoggingAuditTrail` that deal with the common preamble fields of every log entry (such as the host address and the node id).

We need to make this a base class which exposes a log entry builder that implementors can populate with values for their respective event types.

Reply:
> Can we just pass LoggingAuditTrail.entryCommonFields.commonFields at instantiation time of LogEntryBuilder?

The `LogEntryBuilder` constructor does it already.
```
        LogEntryBuilder(boolean showOrigin) {
            logEntry = new StringMapMessage(LoggingAuditTrail.this.entryCommonFields.commonFields);
            if (false == showOrigin) {
                logEntry.remove(ORIGIN_ADDRESS_FIELD_NAME);
                logEntry.remove(ORIGIN_TYPE_FIELD_NAME);
            }
        }
```
_(GH doesn't allow links in PRs, unless you comment at the place)._

Do you mean making the log entry builder class and the common fields protected, and using them outside the `LoggingAuditTrail` class?
The solution is indeed to separate the `AuditTrail` implementation(s) from a supplier of log entry builders, but the devil is in the details.

I am happy to prototype your proposal, but I need more information:
Where would the log entry builder be called from? How do we instantiate the log entry builder? How would this separate the tests? Maybe we need to discuss this in a conversation.

I can prototype the separation, on the existing code or the code in this PR, but do you think we can deal with it after this is merged?

Reply:
What I mean is to have:
```java
        LogEntryBuilder(Map<String, String> commonFields, boolean includeRequestBody, boolean showOrigin) {
            logEntry = new StringMapMessage(commonFields);
            ...
        }
```
So the existing call sites will updated to:
```java
new LogEntryBuilder(entryCommonFields.commonFields, false, false)
   ...
```

I believe at this point, the class can be extracted into a its own file. The tests can be separated broadly to test just the `LogEntryBuilder` and the `LoggintAuditTrail`. 

Tests for `LogEntryBuilder` are mainly for the message format, i.e. tests for the different overloaded versions of the `withRequestBody` method, which does not require a `LoggintAuditTrail`.

Tests for `LoggintAuditTrail` can just ensure relevant methods of the `LogEntryBuilder` is called without worrying about the format of the final string message. It would be great if we can just mock `LogEntryBuilder` so its easier to test. An option could be passing a supplier to `LoggintAuditTrail`, e.g.:
```java
public LoggintAuditTrail(Supplier<LogEntryBuilder> builderSupplier, ...) {
   ...
}
```
And uses it like:
```java
builderSupplier.get().with(...).with(...)
```

I hope this makes sense. I have no issue to defer these refactoring after this PR is merged.

Reply:
Thank you for the suggestion.

In principle, I agree we can separate the log building class more. But the constructor that you're proposing takes a map with obscure values that I don't think we should expose publicly. I also don't think testing the `StringMapMessage`'s contents and then testing that the contents map to the actual output logfile improves on the testing we currently have.

Reply:
Since this is such a debated topic, I'll make sure we'll sit down and discuss it before I start working on the refactoring.